File: src/0589.n-ary-tree-preorder-traversal.775/solution.ts
class Node1 {
    val: number
    children: Node1[]
    constructor (val?: number) {
        this.val = (val === undefined ? 0 : val);
        this.children = [];
    }
}

function preorder (root: Node1 | null): number[] {
    if (!root) {
        return [];
    }
    const stack:Node1[] = [root, ];
    const result:number[] = [];
    while (stack.length) {
        const node = stack.pop()!;
        result.push(node.val);
        for (let i = node.children.length - 1; i > -1; i--) {
            stack.push(node.children[i]);
        }
    }
    return result;
}

File: src/0589.n-ary-tree-preorder-traversal.775/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> result = new ArrayList<>();
        dfs(root,result);
        return result;
    }
    
    private void dfs(Node root,List<Integer> result){
        if(root == null){
            return;
        }
        result.add(root.val);
        
        for(Node node : root.children){
            dfs(node,result);
        }
        
    }
    
}